src/mini_batch.jl

```diff
@@ -26,99 +33,100 @@ function kmeans!(alg::MiniBatch, X, k;
     N = zeros(T, k)
 
     # Initialize nearest centers for both batch and whole dataset labels
-    final_labels = Vector{Int}(undef, ncol)  # dataset labels
-
     converged = false
     niters = 0
     counter = 0
     J_previous = zero(T)
     J = zero(T)
+    totalcost = zero(T)
 
-    # TODO: Main Steps. Batch update centroids until convergence
+    # Main Steps. Batch update centroids until convergence
     while niters <= max_iters  # Step 4 in paper
 
         # b examples picked randomly from X (Step 5 in paper)
         batch_rand_idx = isnothing(weights) ? rand(rng, 1:ncol, alg.b) : wsample(rng, 1:ncol, weights, alg.b)
-        batch_sample = X[:, batch_rand_idx]
 
         # Cache/label the batch samples nearest to the centers (Step 6 & 7)
-        @inbounds for i in axes(batch_sample, 2)
-            min_dist = distance(metric, batch_sample, centroids, i, 1)
+        @inbounds for i in batch_rand_idx
+            min_dist = distance(metric, X, centroids, i, 1)
             label = 1
 
             for j in 2:size(centroids, 2)
-                dist = distance(metric, batch_sample, centroids, i, j)
+                dist = distance(metric, X, centroids, i, j)
                 label = dist < min_dist ? j : label
                 min_dist = dist < min_dist ? dist : min_dist
             end
 
-            final_labels[batch_rand_idx[i]] = label
+            containers.labels[i] = label
         end
 
-        # TODO: Batch gradient step
-        @inbounds for j in axes(batch_sample, 2)  # iterate over examples (Step 9)
+        # Batch gradient step
+        @inbounds for j in batch_rand_idx  # iterate over examples (Step 9)
 
-            # Get cached center/label for this x  => labels[batch_rand_idx[j]] (Step 10)
-            label = final_labels[batch_rand_idx[j]]
+            # Get cached center/label for this x  => (Step 10)
+            label = containers.labels[j]
+            
             # Update per-center counts
-            N[label] += isnothing(weights) ? 1 : weights[j]  # verify (Step 11)
+            N[label] += isnothing(weights) ? 1 : weights[j]  # (Step 11)
 
             # Get per-center learning rate (Step 12)
             lr = 1 / N[label]
 
-            # Take gradient step (Step 13) # TODO: Replace with an allocation-less loop.
-            centroids[:, label] .= (1 - lr) .* centroids[:, label] .+ (lr .* batch_sample[:, j])
+            # Take gradient step (Step 13) # TODO: Replace with faster loop?
+            @views centroids[:, label] .= (1 - lr) .* centroids[:, label] .+ (lr .* X[:, j])
         end
 
-        # TODO: Reassign all labels based on new centres generated from the latest sample
-        final_labels = reassign_labels(X, metric, final_labels, centroids)
+        # Reassign all labels based on new centres generated from the latest sample
+        containers.labels .= reassign_labels(X, metric, containers.labels, centroids)
```

Comment:
As we discussed in Zulip, there is no `reassign` function in the definition of the algorithm. We are using it only to build further `J` which in turn is used for early stopping. In this implementation, we introduce a huge slowdown by using full labels recalculation and thus neglecting all benefits of mini-batch algorithm.

Reply:
Fully agree. Need a better way to do early stopping from the batch costs.